packages/smithy-aws-core/src/smithy_aws_core/credentials_resolvers/imds.py

```diff
@@ -0,0 +1,230 @@
+#  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#  SPDX-License-Identifier: Apache-2.0
+import json
+import threading
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import Literal
+
+from smithy_core import URI
+from smithy_core.aio.interfaces.identity import IdentityResolver
+from smithy_core.exceptions import SmithyIdentityException
+from smithy_core.interfaces.identity import IdentityProperties
+from smithy_core.interfaces.retries import RetryStrategy
+from smithy_core.retries import SimpleRetryStrategy
+from smithy_http import Field, Fields
+from smithy_http.aio import HTTPRequest
+from smithy_http.aio.interfaces import HTTPClient
+
+from smithy_aws_core.identity import AWSCredentialsIdentity
+
+
+class Token:
+    """Represents an IMDSv2 session token with a value and method for checking
+    expiration."""
+
+    def __init__(self, value: bytes, ttl: int):
+        self._value = value
+        self._ttl = ttl
+        self._created_time = datetime.now()
+
+    def is_expired(self) -> bool:
+        """Check if the token has expired."""
+        return datetime.now() - self._created_time >= timedelta(seconds=self._ttl)
+
+    @property
+    def value(self) -> bytes:
+        return self._value
+
+
+class TokenCache:
+    """Holds the token needed to fetch instance metadata. In addition, it knows how to
+    refresh itself.
+
+    :param HTTPClient http_client: The client used for making http requests.
+    :param int token_ttl: The time in seconds before a token expires.
+    """
+
+    _MIN_TTL = 5
+    _MAX_TTL = 21600
+    _TOKEN_PATH = "/latest/api/token"
+
+    def __init__(
+        self, http_client: HTTPClient, base_uri: URI, token_ttl: int = _MAX_TTL
+    ):
+        self._http_client = http_client
+        self._base_uri = base_uri
+        self._token_ttl = self._validate_token_ttl(token_ttl)
+        self._refresh_lock = threading.Lock()
```

Comment:
This may be my lack of python/async knowledge, but can we still use a threading.Lock with async?  I think we might instead need asyncio.Lock here?

Reply:
Yup, this was a miss when porting over some of the sync code we'd written in the past for this.

Fixed in a585f5de3d783a2ab1b6af5ee506c4975ba74e28